protocol-models/src/main/resources/airbyte_protocol/protocol_methods.md

```diff
@@ -0,0 +1,28 @@
+# Protocol Methods
+
+The purpose of this doc is to define the interface on the docker containers that implement the Airbyte Protocol. It is purposefully lean, as we describe the protocol at length in [our docs](https://docs.airbyte.com/understanding-airbyte/airbyte-protocol). We just want to define the interface here, so that as it evolves, we can version the change as part of the protocol version.
+
+## Method Interfaces
+
+We describe these interfaces in pseudocode for clarity. Clarifications on the pseudocode semantics:
+* Any `Stream~ that is mentioned as input arg, is passed to the docker contained via STDIN. 
+* All other parameters are passed in as command line args (e.g. --config <path to config file>). 
+* Each input parameter is described as its type (as defined in airbyte_protocol.yml and the name of the parameter). 
+* If an input parameter has no name, then it is passed via STDIN.
+
+In addition to the return types mentioned below, all methods can return the following message types: `AirbyteLogMessage | AirbyteTraceMessage`.
+
+**interface of both source and destination**
+```
+spec() -> Stream<AirbyteConnectorSpecification>
+check(Config, Optional<ConfiguredCatalog>) -> Stream<AirbyteConnectionStatus>
+```
+**source only**
+```
+discover(Config) -> AirbyteCatalog
+read(Config, ConfiguredAirbyteCatalog, State) -> Stream<AirbyteRecordMessage | AirbyteStateMessage | AirbyteControlMessage>
```

Comment:
I think this method-style representation is a bit misleading.  The flag argument isn't `--ConfiguredAirbyteCatalog` it is `--catalog`.  We also don't explain if are passing the object itself (e.g. stringified JSON) or a file path.

Reply:
Perhaps a way to keep this method-like signature could be:
```
read(config -> File<Config>, catalog -> File<ConfiguredAirbyteCatalog>, state -> File<State>) -> Stream<...>
```

... but now I'm just making things up.

I *think* JSONSchema for this might work better for this:

```js
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "additionalProperties": true // this is important now
  "required": ["config", "catalog"] // showing that state is optional
  "arguments": {
    "config": { "type": "file_path", "$ref": Config.yaml },
    "catalog": { "type": "file_path", "$ref": ConfiguredAirbyteCatalog.yaml },
    "state": { "type": "file_path", "$ref": State.yaml },
  }
}

```

Reply:
Here's a real READ command for reference:

```
docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/source-faker:dev read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json
```

Reply:
I do think the JSON schema version is a little clearer, maybe there's a way to designate STDIN/STDOUT parameters from arguments to the method call?